### PR TITLE
NLL: Walk the MIR only once for the "unused mut" lint

### LINKED ISSUE #51641: [NLL] used-mut lint has an O(n^2) loop
The following snippet of code is an O(n^2) loop:

https://gist.github.com/lqd/7e178cb9ac8302e24272965048627f21#file-mod-rs-L243-L256

The problem is that we are iterating over each temporary and then invoking `find_assignments`, which basically walks over the MIR again. What we really want to be doing is walking the MIR at most once (ideally, coallesced into some other walk) and registered the "used mut" assignments.

### PR DESCRIPTION
**Pull request by lqd**
Turns the quadratic loop gathering local variable assignments into a single MIR walk, and brings down the number of `super_mir` calls generated from `do_mir_borrowck` to the expected levels seen in `nll::replace_regions_in_mir` and `nll::compute_regions`, i.e. on clap: 1883 `super_mir` calls instead of 8011.

The limited perf numbers I could gather on my machines look to be what we expected: `clap-check` seems to be gaining back a lot of the 7% we previously saw in `visit_mir`.

Fixes #51641.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_mir/borrow_check/used_muts.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        local: &Local,
+        place_context: PlaceContext<'tcx>,
+        location: Location,
+    ) {
+        if !self.needles.contains(local) {
+            return;
+        }
+
+        match place_context {
+            PlaceContext::Store | PlaceContext::Call => {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/used_muts.rs]: I wonder if we should factor this out to share code with the `find_assignments`

Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/used_muts.rs]: basically just make a fn that returns a `bool`

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/librustc_mir/borrow_check/used_muts.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/used_muts.rs]: maybe share code with this function?

https://github.com/rust-lang/rust/blob/637fd2e0487281adac99000602f5e74c3bf151a8/src/librustc_mir/util/liveness.rs#L229-L229

**Comment by kennytm** [comment]
Perf is ready. http://perf.rust-lang.org/compare.html?start=637fd2e0487281adac99000602f5e74c3bf151a8&end=737c5cc95040e9efce65cb13a51a0df0f842cef4&stat=instructions%3Au

**Comment by lqd** [comment]
@kennytm the failures look like CI network issues right ?

**Pull request merged**